### ISSUE TITLE
Compilation error when using *ngFor to iterate over unions of arrays of arrays of string literals

### ISSUE THREAD

**Issue by zilia-gmethot** [angular/angular#49310]
### Which @angular/* package(s) are the source of the bug?

core

### Is this a regression?

No

### Description

Using the *ngFor directive, It is impossible to iterate over a variable of type union of arrays of arrays of string literals without obtaining a compilation error, even though it is valid TypeScript.

### Please provide a link to a minimal reproduction of the bug

https://stackblitz.com/edit/angular-7hhamp?file=src/main.ts

### Please provide the exception or error you saw

```true
Error in src/main.ts (16:36)
Type 'FooBar' is not assignable to type 'NgIterable<Foo[]>'.
```

### Please provide the environment you discovered this bug in (run `ng version`)

```true
Angular CLI: 12.2.18
Node: 16.19.0 (Unsupported)
Package Manager: npm 8.19.3
OS: linux x64

Angular: 12.2.17
... animations, common, compiler, compiler-cli, core, forms
... language-service, platform-browser, platform-browser-dynamic
... router

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1202.18
@angular-devkit/build-angular   12.2.18
@angular-devkit/core            12.2.18
@angular-devkit/schematics      12.2.18
@angular/cdk                    12.2.13
@angular/cli                    12.2.18
@angular/material               12.2.13
@schematics/angular             12.2.18
rxjs                            6.6.7
typescript
```

**Comment by MaximPVladimirovich** [comment]
You are getting this error because your type FooBar is not compatible with the type NgIterable<Foo[]> that is expected by Angular's *ngFor directive. According to the Angular documentation](https://angular.io/api/core/NgIterable), NgIterable is defined as type NgIterable<T> = Array<T> | Iterable<T>;.  

Changing line 24 to this seems to do the trick. 
`values: NgIterable<FooBar> = [['a'], ['e'], ['c']]`

 and changing 
`type FooBar = Foo[][] | Bar[][]` to `type FooBar = Foo[] | Bar[];`

**Comment by zilia-gmethot** [comment]
> You are getting this error because your type FooBar is not compatible with the type NgIterable<Foo[]> that is expected by Angular's *ngFor directive. According to the Angular documentation](https://angular.io/api/core/NgIterable), NgIterable is defined as type NgIterable = Array | Iterable;.


You are mistaken. Given the following type definition from the Angular documentation:
```
type NgIterable<T> = Array<T> | Iterable<T>;
```

Then the type `Foo[][]`, which is equivalent to `Array<Foo[]>`, satisfies `NgIterable<Foo[]>`. The following example, which compiles just fine using the TypeScript compiler, proves it:
```
type NgIterable<T> = Array<T> | Iterable<T>;
type Foo = 'foo';
type Bar = 'bar';
type FooBar = Foo[][] | Bar[][];

function f(x: NgIterable<Foo[]>): void {
    console.log(x);
}

const values: FooBar = [['foo']];

f(values);
```

> Changing line 24 to this seems to do the trick. `values: NgIterable<FooBar> = [['a'], ['e'], ['c']]`

For my use case, this value should not be valid.

> and changing `type FooBar = Foo[][] | Bar[][]` to `type FooBar = Foo[] | Bar[];`

I can't simply change two-dimensional arrays to one-dimensional arrays. My use case requires two-dimensional arrays.

**Comment by briosheje** [comment]
The issue seems to be somewhat related to any union type, the easiest reproducible example can be found here: https://stackblitz.com/edit/angular-qndby2?file=src%2Fmain.ts inspired from this stackoverflow question: https://stackoverflow.com/questions/75662030/why-doesnt-ngforof-directive-in-angular-work-with-union-types

It looks like that any union type is somehow ignored and only the first type will be taken for some reasons.

For example, this example:
```
import 'zone.js/dist/zone';
import { Component } from '@angular/core';
import { CommonModule } from '@angular/common';
import { bootstrapApplication } from '@angular/platform-browser';

@Component({
  selector: 'my-app',
  standalone: true,
  imports: [CommonModule],
  template: `
    <h1>Hello from {{name}}!</h1>
    <a target="_blank" href="https://angular.io/start">
      Learn more about Angular 
    </a>
    <div *ngFor="let element of array">
      {{element}}
    </div>
  `,
})
export class App {
  name = 'Angular';

  array: string[] | number[] = ['1', '2'];
}

bootstrapApplication(App);
```

Will produce the following error: `Type 'string[] | number[]' is not assignable to type 'NgIterable<string>'.`

While this example (swapping string[] and number[]):

```
import 'zone.js/dist/zone';
import { Component } from '@angular/core';
import { CommonModule } from '@angular/common';
import { bootstrapApplication } from '@angular/platform-browser';

@Component({
  selector: 'my-app',
  standalone: true,
  imports: [CommonModule],
  template: `
    <h1>Hello from {{name}}!</h1>
    <a target="_blank" href="https://angular.io/start">
      Learn more about Angular 
    </a>
    <div *ngFor="let element of array">
      {{element}}
    </div>
  `,
})
export class App {
  name = 'Angular';

  array: number[] | string[] = ['1', '2'];
}

bootstrapApplication(App);
```

Will produce the following error: `Type 'number[] | string[]' is not assignable to type 'NgIterable<number>'.`

**Comment by demirmusa** [comment]
Hi @briosheje, you are right. I encountered the same error, I tested it in a few cases and got the same result.

Using array instead might be a workaround for people encountering the same error.

```diff
export class App {
  name = 'Angular';

-  array: number[] | string[] = ['1', '2'];
+  array: Array<number | string> = ['1', '2'];
}
```

then it works

```html
<div *ngFor="let element of array">
  {{element}}
</div>
```

**Comment by zilia-gmethot** [comment]
> Hi @briosheje, you are right. I encountered the same error, I tested it in a few cases and got the same result.
> 
> Using array instead might be a workaround for people encountering the same error.
> 
> ```diff
> export class App {
>   name = 'Angular';
> 
> -  array: number[] | string[] = ['1', '2'];
> +  array: Array<number | string> = ['1', '2'];
> }
> ```
> 
> then it works
> 
> ```
> <div *ngFor="let element of array">
>   {{element}}
> </div>
> ```

This is not an acceptable workaround because it is not the same type. You do not want the array to possibly contain both strings and numbers.

```
/**
 * @param x - An array containing either a mix of strings and numbers, only strings or only numbers.
 */
function a(x: (string | number)[]): void {
  return;
}

/**
 * @param x - Either an array containing only strings, or an array containing only numbers.
 */
function b(x: string[] | number[]): void {
  return;
}

// Valid examples
a(['a', 1]);
a(['a', 'b']);
a([1, 1]);
b(['a', 'b']);
b([1, 2]);

// Invalid example
b(['a', 1]);
```

**Comment by demirmusa** [comment]
Yes, you are right, it is not the same thing. You can only use it to avoid errors if you wrote the code and you already know that you will use array as string[] or number[]. But if someone else is going to use the code, you are allowing them to send data in a way you do not want. Of course, it's not the right solution, but it's just a workaround that can prevent you from getting errors in certain situations.

**Comment by xbisa** [comment]
There's a solution that is not a workaround: **use User-defined Type Guards.** I have a flexible API endpoint that returns `User[] | string[]`

```
export function isUser(user: User | string | null): user is User {
  return user !== null && !!(user as User)?.id;
}

export function isUserArray(users: User[] | string[]): users is User[] {
  const [ user ] = users;
  return isUser(user);
}
```
```
<ng-container *ngIf='isUserArray(user.coaches)'>
</ng-container>
```

**Comment by JeanMeche** [comment]
This now works fine with the `@for` block, and since we've deprecated the `NgFor` structural directive, we won't be addressing this issue.
